### PR TITLE
[hail] automatically build two bioinformatics images 

### PR DESCRIPTION
**Pull request by danking**
This adds two new Dockerfiles. The first has gsutil and pip-wheel-installed
Hail. The second builds on the first adding a number of bioinformatics tools
that were included in the CCG Tutorial.

Deployment to dockerhub is not trivial because, unfortunately, I need to mount
the docker socket even to download and then upload an image (never starting a
container). I'll design and implement some extension to CI that lets me deploy
images to docker hub later. For now, I used dev deploy to build
these (https://ci.hail.is/batches/33294) and then manually pulled them and
uploaded them to PyPI as hailgenetics/hail:0.2.37 and
hailgenetics/genetics:0.2.37. I particularly appreciate criticism of the names.

### CODE REVIEW

File: docker/hail/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      matplotlib \
+      numpy \
+      pandas \
+      scikit-learn \
+      scipy \
+    && rm -rf hail-*-py3-none-any.whl
+RUN curl https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-hadoop2-2.0.1.jar \
+         > $(find_spark_home.py)/jars/gcs-connector-hadoop2-2.0.1.jar && \
+    mkdir -p $(find_spark_home.py)/conf && \
+    touch $(find_spark_home.py)/spark-defaults.conf && \
```
Comment (jigold) [inline docker/hail/Dockerfile:32]: does this file need to be in the conf subdirectory you created above?

File: docker/genetics/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+     curl -LO https://github.com/samtools/samtools/releases/download/1.9/samtools-1.9.tar.bz2 && \
+     tar -xf samtools-1.9.tar.bz2 && \
+     rm -rf samtools-1.9.tar.bz2 && \
+     cd samtools-1.9 && \
+     ./configure --without-curses && \
+     make && \
+     make install)
+RUN git clone --recursive https://github.com/vcflib/vcflib.git && \
+    (cd vcflib && make)
+RUN curl -LO http://s3.amazonaws.com/plink2-assets/plink2_linux_avx2_latest.zip && \
```
Comment (jigold) [inline docker/genetics/Dockerfile:23]: What's your reasoning for using latest? I would think we'd want to version control what the contents of each image are so someone could cite what version they used.

File: docker/Makefile [hunk 3]
```diff
@@ -21,6 +21,18 @@ service-base:
 	python3 ../ci/jinja2_render.py '{"base_image":{"image":"base"}}' Dockerfile.service-base Dockerfile.service-base.out
 	docker build -t service-base -f Dockerfile.service-base.out --cache-from service-base,$(SERVICE_BASE_LATEST),base ..
 
+.PHONY: hail-public-image
+hail-public-image:
+	make -C ../hail wheel
+	cp ../hail/build/deploy/dist/hail-*-py3-none-any.whl hail/
+	cd hail && tar -cvf wheel-container.tar hail-*-py3-none-any.whl
+	docker build hail -f hail/Dockerfile -t hail-public-image
```
Comment (jigold) [inline docker/Makefile:29]: We probably want to use the cache for ubuntu 18.04 and the latest version of the hail image.

File: docker/Makefile [hunk 4]
```diff
@@ -21,6 +21,18 @@ service-base:
...
+
+.PHONY: genetics-public-image
+genetics-public-image: hail-public-image
+	python3 ../ci/jinja2_render.py '{"hail_public_image":{"image":"hail-public-image"}}' genetics/Dockerfile genetics/Dockerfile.out
+	docker build genetics -f genetics/Dockerfile.out -t genetics-public-image
```
Comment (jigold) [inline docker/Makefile:34]: Same question about using the cache.

File: docker/Makefile (hunk 3 above)
Comment (jigold) [inline docker/Makefile:29]: do we tag the hail-public-image with the version like 0.2.37?

File: build.yaml [hunk 5]
```diff
@@ -1551,3 +1551,42 @@ steps:
     - copy_files
     - ci_utils_image
     - build_hail
+ - kind: buildImage
+   name: hail_public_image
+   dockerFile: docker/hail/Dockerfile
+   contextPath: docker/hail/
+   publishAs: hail_public
+   dependsOn:
+    - build_hail
```
Comment (jigold) [inline build.yaml:1560]: does this work now? I thought you had to run dev deploy for it to work. I was expecting to see scopes set to dev only.

Comment (danking) [reply build.yaml:1560]: The image build works fine, I just have no means to upload the resulting images to DockerHub.

As normal build steps, they are tested regularly and built on each master merge. Updating DockerHub is as easy as loading the deploy build and finding the gcr.io name for the latest built image.

File: docker/Makefile (hunk 3 above)
Comment (danking) [reply docker/Makefile:29]: The images built by CI are tagged with the SHA. When I manually upload to DockerHub I do tag with the latest hail version.

File: docker/genetics/Dockerfile (hunk 2 above)
Comment (danking) [reply docker/genetics/Dockerfile:23]: Chris Chang deletes old versions when he posts new versions so we'd have to PR a new URL every couple weeks to fix failing deploys.

Comment (danking) [reply docker/genetics/Dockerfile:23]: And I kind of trust plink to not break something 🤷

File: docker/hail/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      hail-*-py3-none-any.whl \
+      ipython \
+      matplotlib \
+      numpy \
+      pandas \
+      scikit-learn \
+      scipy \
+    && rm -rf hail-*-py3-none-any.whl
+RUN curl https://storage.googleapis.com/hadoop-lib/gcs/gcs-connector-hadoop2-2.0.1.jar \
+         > $(find_spark_home.py)/jars/gcs-connector-hadoop2-2.0.1.jar && \
```
[the remark this answers is not in the document]
Comment (danking) [reply docker/hail/Dockerfile:30]: hail depends on pyspark which includes `find_spark_home.py`.

File: docker/hail/Dockerfile (hunk 1 above)
Comment (danking) [reply docker/hail/Dockerfile:32]: sure does. fixed

File: docker/hail/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    mkdir -p $(find_spark_home.py)/conf && \
+    touch $(find_spark_home.py)/spark-defaults.conf && \
+    sed -i $(find_spark_home.py)/spark-defaults.conf \
+        -e 's:spark\.hadoop\.google\.cloud\.auth\.service\.account\.enable.*:spark.hadoop.google.cloud.auth.service.account.enable true:' \
+        -e 's:spark\.hadoop\.google\.cloud\.auth\.service\.account\.json\.keyfile.*:spark\.hadoop\.google\.cloud\.auth\.service\.account\.json\.keyfile /gsa-key/key.json:'
```
[the remark this answers is not in the document]
Comment (danking) [reply docker/hail/Dockerfile:35]: It only works out-of-the-box with batch, but a user can modify this file or run Ben Weisburd's http://broad.io/install-gcs-connector .

Getting google credentials into a docker container is kind of annoying and weird in any case, I don't expect non-sophisticated users to do it outside of Batch.

**Pull request merged**